hydro_deploy/core/src/util.rs

```diff
@@ -26,60 +26,73 @@ pub async fn async_retry<T, F: Future<Output = Result<T>>>(
 }
 
 type PriorityBroadcacst = (
-    Arc<RwLock<Option<tokio::sync::oneshot::Sender<String>>>>,
-    Arc<RwLock<Vec<Sender<String>>>>,
+    Arc<Mutex<Option<tokio::sync::oneshot::Sender<String>>>>,
+    Arc<Mutex<Vec<Sender<String>>>>,
 );
 
 pub fn prioritized_broadcast<T: Stream<Item = io::Result<String>> + Send + Unpin + 'static>(
     mut lines: T,
     default: impl Fn(String) + Send + 'static,
 ) -> PriorityBroadcacst {
-    let priority_receivers = Arc::new(RwLock::new(None::<tokio::sync::oneshot::Sender<String>>));
-    let receivers = Arc::new(RwLock::new(Vec::<Sender<String>>::new()));
+    let priority_receivers = Arc::new(Mutex::new(None::<tokio::sync::oneshot::Sender<String>>));
+    let receivers = Arc::new(Mutex::new(Vec::<Sender<String>>::new()));
 
     let weak_priority_receivers = Arc::downgrade(&priority_receivers);
     let weak_receivers = Arc::downgrade(&receivers);
 
     tokio::spawn(async move {
-        'line_loop: while let Some(Result::Ok(line)) = lines.next().await {
+        while let Some(Result::Ok(line)) = lines.next().await {
             if let Some(cli_receivers) = weak_priority_receivers.upgrade() {
-                let mut cli_receivers = cli_receivers.write().await;
+                let mut cli_receivers = cli_receivers.lock().unwrap();
 
                 let successful_send = if let Some(r) = cli_receivers.take() {
                     r.send(line.clone()).is_ok()
                 } else {
                     false
                 };
+                drop(cli_receivers);
 
                 if successful_send {
-                    continue 'line_loop;
+                    continue;
                 }
             }
 
             if let Some(receivers) = weak_receivers.upgrade() {
-                let mut receivers = receivers.write().await;
-                let mut successful_send = false;
-                for r in receivers.iter() {
-                    successful_send |= r.send(line.clone()).await.is_ok();
-                }
-
-                receivers.retain(|r| !r.is_closed());
+                let send_tasks = {
```

Comment:
This seems like a lot of indirection just to change an async lock into a sync lock. Maybe just have `receivers` use the original `tokio::RwLock` to simplify the code here? Also is unfortunate that we have to clone the senders. I do think `join_all` is good, though, so if we can replace the for loop with that, it's a win.

Reply:
Cloning a `receiver: Sender<String>` is just an `Arc` increment though

Reply:
Oh I do have an extra clone that should be removed though

Reply:
If I use tokio::mpsc::unbounded_channel instead of async_channel then the method doesn't need to be async at all - is there any particular reason you chose to use `async-channel?`

Reply:
I guess `Receiver` only needs `&self` to receive, while `UnboundedReceiver` needs `&mut self`

Reply:
Ooh yeah we could use `tokio` and for the Python side we can totally just use a `RefCell`.